src/ports/postgres/modules/utilities/pivot.py_in

```diff
@@ -58,66 +61,255 @@ def pivot(schema_madlib, source_table, out_table,
         FROM pivset GROUP BY id ORDER BY id)
 
     """
+
     indices = split_quoted_delimited_str(index)
-    pcol = split_quoted_delimited_str(pivot_cols)
-    pval = split_quoted_delimited_str(pivot_values)
-    validate_pivot_coding(source_table, out_table, indices, pcol, pval)
-    new_col_names =[]
-    sql_list = ["CREATE TABLE " + out_table + " AS (SELECT " + index]
+    pcols = split_quoted_delimited_str(pivot_cols)
+    pvals = split_quoted_delimited_str(pivot_values)
+    validate_pivot_coding(source_table, out_table, indices, pcols, pvals)
 
-    pcol_no_quotes = strip_end_quotes(pcol[0].strip())
-    pval_no_quotes = strip_end_quotes(pval[0].strip())
+    # Strip end quotes from pivot columns
+    pcols = [strip_end_quotes(pcol.strip()) for pcol in pcols]
+    pvals = [strip_end_quotes(pval.strip()) for pval in pvals]
+
+    # Parse the aggregate_func as a dictionary
+    try:
+        param_types = dict.fromkeys(pvals, list)
+        param_types['__madlib_def_pval__'] = list
+        agg_dict = extract_keyvalue_params(aggregate_func,param_types)
+    except KeyError, e:
+        with MinWarning("warning"):
+            plpy.warning("Pivot: Not all columns from '{aggregate_func}'"
+            " present in '{pivot_values}'".format(**locals()))
+            raise
+
+    # If the dictionary is empty, parse it as a list
+    agg_set = split_quoted_delimited_str(aggregate_func) if len(agg_dict) < 1 \
+         else []
+    # If the list is empty set 'avg' as default
+    agg_dict['__madlib_def_agg__'] = ['avg'] if len(agg_set) < 1 else agg_set
 
     # Find the distinct values of pivot_cols
-    distinct_values = plpy.execute(
-        "SELECT array_agg(DISTINCT {pcol} ORDER BY {pcol}) AS value "
-        "FROM {source_table}".
-        format(pcol=pcol[0], source_table=source_table))
-
-    distinct_values = [strip_end_quotes(item)
-                        for item in distinct_values[0]['value']]
-    # The aggregate collects pivot_values values for a given pivot_cols value
-    case_str = ("{agg}("
-                "CASE WHEN \"{{pcol}}\" = '{{value}}' THEN {pval} ELSE NULL END"
-                ")".
-                format(agg=aggregate_func,
-                       pval=pval_no_quotes))
-    sql_list.append(
-        ", " +
-        # Assign the name of the new column
-        ', '.join("{case_str} as \"{{pcol}}_{{value}}\"".
-                  format(case_str=case_str).
-                  format(pcol=pcol_no_quotes, value=str(value))
-                  for value in distinct_values if value is not None))
+    # Note that the distinct values are appended in order of users list
+    # This ordering is important when we access pivot_comb entries
+    distinct_str = ["SELECT "]
+    distinct_str.append(
+        ', '.join("array_agg(DISTINCT {pcol}) AS {pcol}_values".
+            format(pcol=pcol) for pcol in pcols))
+    distinct_str.append(" FROM " + source_table)
+    distinct_values = plpy.execute(''.join(distinct_str))
```

Comment:
Why not create a single query string instead of joining them? 
```
array_agg_str = ', '.join("array_agg(DISTINCT {pcol}) AS {pcol}_values".format(pcol=pcol) for pcol in pcols)
plpy.execute("SELECT {0} FROM {1}".format(array_agg_str, source_table))
```